File: platform/polyglot/github_api/py/github_api/sdk/rate_limit.py
"""
Rate limit parsing, detection, and wait logic for the GitHub API.

Handles both primary rate limits (x-ratelimit-* headers) and
secondary rate limits (403/429 with Retry-After).
"""

from __future__ import annotations

import asyncio
import logging
from datetime import datetime, timezone
from typing import Any, Mapping

from pydantic import BaseModel, Field

__all__ = [
    "RateLimitInfo",
    "parse_rate_limit_headers",
    "should_wait_for_rate_limit",
    "wait_for_rate_limit",
    "is_secondary_rate_limit",
]


class RateLimitInfo(BaseModel):
    """Parsed rate limit information from GitHub API response headers."""

    limit: int = Field(description="Maximum number of requests per window")
    remaining: int = Field(description="Requests remaining in current window")
    reset: int = Field(description="Unix timestamp when the window resets")
    used: int = Field(default=0, description="Requests used in current window")
    resource: str = Field(default="core", description="Rate limit resource category")

    @property
    def reset_at(self) -> datetime:
        """Return reset time as a datetime object (UTC)."""
        return datetime.fromtimestamp(self.reset, tz=timezone.utc)

    @property
    def seconds_until_reset(self) -> float:
        """Seconds until the rate limit window resets."""
        now = datetime.now(tz=timezone.utc).timestamp()
        return max(0, self.reset - now)

    @property
    def is_exhausted(self) -> bool:
        """Whether the rate limit is exhausted."""
        return self.remaining == 0


def parse_rate_limit_headers(
    headers: Mapping[str, str],
) -> RateLimitInfo | None:
    """Parse rate limit info from GitHub API response headers.

    Args:
        headers: HTTP response headers (case-insensitive mapping).

    Returns:
        RateLimitInfo if rate limit headers are present, None otherwise.
    """
    limit_str = headers.get("x-ratelimit-limit")
    remaining_str = headers.get("x-ratelimit-remaining")
    reset_str = headers.get("x-ratelimit-reset")

    if limit_str is None or remaining_str is None or reset_str is None:
        return None

    try:
        limit = int(limit_str)
        remaining = int(remaining_str)
        reset = int(reset_str)
    except (ValueError, TypeError):
        return None

    used_str = headers.get("x-ratelimit-used", "0")
    try:
        used = int(used_str)
    except (ValueError, TypeError):
        used = limit - remaining

    resource = headers.get("x-ratelimit-resource", "core")

    return RateLimitInfo(
        limit=limit,
        remaining=remaining,
        reset=reset,
        used=used,
        resource=resource,
    )


def should_wait_for_rate_limit(
    info: RateLimitInfo,
    *,
    auto_wait: bool = True,
    threshold: int = 0,
) -> bool:
    """Determine if the client should wait before making another request.

    Args:
        info: Current rate limit information.
        auto_wait: Whether auto-wait is enabled.
        threshold: Remaining count threshold below which to wait.

    Returns:
        True if the client should wait.
    """
    if not auto_wait:
        return False
    return info.remaining <= threshold


async def wait_for_rate_limit(
    info: RateLimitInfo,
    logger: logging.Logger | None = None,
) -> None:
    """Sleep until the rate limit window resets.

    Args:
        info: Rate limit information containing the reset timestamp.
        logger: Optional logger for status messages.
    """
    wait_seconds = info.seconds_until_reset + 1  # 1s buffer
    if wait_seconds <= 0:
        return

    if logger:
        logger.warning(
            "Rate limit exhausted. Waiting %.1f seconds until reset at %s.",
            wait_seconds,
            info.reset_at.isoformat(),
        )

    await asyncio.sleep(wait_seconds)


def is_secondary_rate_limit(
    status: int,
    body: dict[str, Any] | None = None,
) -> bool:
    """Detect if a response indicates a secondary (abuse) rate limit.

    Secondary rate limits return 403 or 429 with a message about
    rate limiting or abuse detection.

    Args:
        status: HTTP response status code.
        body: Parsed JSON response body.

    Returns:
        True if the response is a secondary rate limit.
    """
    if status not in (403, 429):
        return False

    if body is None:
        return False

    message = body.get("message", "").lower()
    secondary_indicators = [
        "rate limit",
        "abuse detection",
        "secondary rate",
        "retry later",
        "you have exceeded a secondary rate limit",
    ]

    return any(indicator in message for indicator in secondary_indicators)

File: platform/polyglot/github_api/mjs/src/sdk/rate-limit.mjs
/**
 * Rate limit parsing, detection, and auto-wait utilities.
 * @module sdk/rate-limit
 */

/**
 * @typedef {Object} RateLimitInfo
 * @property {number} limit - Maximum requests per window
 * @property {number} remaining - Remaining requests in current window
 * @property {number} reset - Unix timestamp when the limit resets
 * @property {number} used - Requests used in current window
 * @property {string} resource - Rate limit resource category
 */

/**
 * Parse rate limit information from response headers.
 * @param {Headers|Object} headers - Response headers
 * @returns {RateLimitInfo|null} Parsed rate limit info, or null if headers missing
 */
export function parseRateLimitHeaders(headers) {
  const get = (name) =>
    headers?.get ? headers.get(name) : headers?.[name];

  const limit = get('x-ratelimit-limit');
  const remaining = get('x-ratelimit-remaining');
  const reset = get('x-ratelimit-reset');
  const used = get('x-ratelimit-used');
  const resource = get('x-ratelimit-resource');

  if (limit == null && remaining == null) {
    return null;
  }

  return {
    limit: limit ? parseInt(limit, 10) : 0,
    remaining: remaining ? parseInt(remaining, 10) : 0,
    reset: reset ? parseInt(reset, 10) : 0,
    used: used ? parseInt(used, 10) : 0,
    resource: resource || 'core',
  };
}

/**
 * Determine whether the client should wait for rate limit reset.
 * @param {RateLimitInfo} info - Current rate limit info
 * @param {Object} [options]
 * @param {boolean} [options.autoWait=true] - Whether auto-wait is enabled
 * @param {number} [options.threshold=0] - Remaining count threshold to trigger wait
 * @returns {boolean} True if the client should wait
 */
export function shouldWaitForRateLimit(info, options = {}) {
  const { autoWait = true, threshold = 0 } = options;

  if (!autoWait || !info) {
    return false;
  }

  return info.remaining <= threshold;
}

/**
 * Sleep until the rate limit resets.
 * @param {RateLimitInfo} info - Rate limit info with reset timestamp
 * @param {Object} [logger] - Logger instance
 * @returns {Promise<void>}
 */
export async function waitForRateLimit(info, logger) {
  if (!info || !info.reset) {
    return;
  }

  const now = Math.floor(Date.now() / 1000);
  const waitSeconds = Math.max(info.reset - now + 1, 1);

  if (logger) {
    logger.warn(`Rate limit reached. Waiting ${waitSeconds}s until reset.`, {
      reset: new Date(info.reset * 1000).toISOString(),
      remaining: info.remaining,
    });
  }

  await new Promise((resolve) => setTimeout(resolve, waitSeconds * 1000));
}

/**
 * Detect whether a response represents a secondary rate limit.
 * Secondary rate limits use 403 or 429 with Retry-After header or specific messages.
 * @param {number} status - HTTP status code
 * @param {Object} body - Parsed response body
 * @returns {boolean} True if this is a secondary rate limit
 */
export function isSecondaryRateLimit(status, body) {
  if (status !== 403 && status !== 429) {
    return false;
  }

  const message = body?.message?.toLowerCase() || '';
  return (
    message.includes('secondary rate limit') ||
    message.includes('abuse detection') ||
    message.includes('you have exceeded a secondary rate limit')
  );
}
